367 ising machine simulated bifurcation

djelovina opened the pull request:
ALP implementation of Ising Machine - Simulated Bifurcation

Commit:
Add Ising machine SB implementation and associated test files

Commit:
Refactor Ising machine SB implementation and enhance debugging output in tests

Commit:
Correct first version
Add vector_print function and update test data for Ising machine SB

Commit:
Enhance Ising machine SB implementation with workspace vector initialization and update test cases for energy validation

Commit:
Refactor Ising machine SB implementation to use template types for sign function and update test cases to use integer types for matrices and vectors

Commit:
Update debug output macros in Ising machine SB implementation to use DEBUG_IMSB

Commit:
Suppress unused parameter warnings in Ising machine SB implementation

Commit:
Fix bug, use size() for vectors

djelovina commented:
Relate to https://github.com/Algebraic-Programming/ALP/pull/370#discussion_r2216480549
set with mask is resizing the vector, therefore I use foldl

`foldl< descr_dense >(  y_comp, mask, ring.template getZero< IOType >(), grb::operators::right_assign<bool,IOType,IOType>()
				);`

@anyzelman is there any better option here?

Commit:
fix bug in foldl and eWiseLambda calls  trigered in nonlocking

Commit:
fix mask use

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+				IsingHType v2;
+				apply( v2, v, v, ring.getMultiplicativeOperator() );
+				foldl( sumJ2, v2, ring.getAdditiveOperator() );
+			}, J );
```
@anyzelman I cannot find any Matrix to Scalar reduction which can support this computations

djelovina replied:
There might be a data race here! https://github.com/Algebraic-Programming/ALP/blob/50b1be9a506e137563bb5fac27e1c5aba67e03cb/include/graphblas/reference/blas3.hpp#L3601 might be a better option with temp matrix ..

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+				rc = rc ? rc : grb::foldl< descr_dense >(  
+					y_comp, mask, ring.template getZero< IOType >(), 
+					grb::operators::right_assign<bool,IOType,IOType>()
+				);
```
@anyzelman It is a bit unclear how this works. There might be a bug here, since if I use non-zero value to assign to the output, the result if different, perhaps downcasted ?  (also with  right_assign_if)

djelovina replied:
bool -> IOType

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+				// 		foldl( x_comp[i], static_cast<IOType>(1), grb::operators::min < IOType >() );
+				// 		//x_comp[i] = std::min<IOType>( 1.0, std::max<IOType>( -1.0, x_comp[i] ) ); 
+				// 	}, x_comp
+				// );
```
@anyzelman not sure which version is better here

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+						sol[i] = sign<IOType, IsingHType>(x_comp[i]);
+					}, 
+					sol, x_comp
+				);
```
There might be a more elegant way to write this

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+#ifdef DEBUG_IMSB
+				std::cout << "e: " << e << '\n';
+#endif
+				energies[ iter ] = e;
```
At the moment I used std::vector for outputs

djelovina commented:
I have a general question, operator, monoid and semiring are passed in the argument list, is there any good reason why they are not template arguments? I ask this because I feel that (element-wise) operators like  abs(x)<1 could be more simply defined in a more structured template-base approach.. @anyzelman

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -0,0 +1,332 @@
+#endif
+
+			    // temp ← Jx + h
+				grb::set( temp, ring.template getZero< IOType >() );
+			    rc = rc ? rc : grb::eWiseApply< descr_dense >(
+			        temp, Jx, h, ring.getAdditiveMonoid()
+			    );
```
Remove set and dense descriptor

djelovina commented on tests/smoke/ising_machine_sb.cpp:
```diff
@@ -0,0 +1,150 @@
+using JType = int;
+
+
+int main( int argc, char ** argv ) {
```
Call from `grbLauncher `

Commit:
Fix right_assign type

Commit:
Use tem J2 matrix to calclate sqared marrix values

Commit:
Refactor test data structure and improve input/output handling in ising_machine_sb.cpp

Commit:
Add iteration tracking and I/O handling in Ising machine tests

Commit:
Refactor test data structure and enhance I/O handling in ising_machine_sb.cpp

Commit:
Enhance is_complex class to support arithmetic types and update Ising machine test to include verification against reference solutions

Commit:
Remove unused example from CMakeLists.txt

Commit:
Refactor is_complex class to improve code formatting and readability

anyzelman commented:
> I have a general question, operator, monoid and semiring are passed in the argument list, is there any good reason why they are not template arguments? I ask this because I feel that (element-wise) operators like abs(x)<1 could be more simply defined in a more structured template-base approach.. @anyzelman

We had this a long long time ago, but found it was easier to program if we could instantiate these algebraic structures and just pass them around, leaving it to C++ to infer the abstract (algebraic) types. I'm not sure if I get your comment re abs though, could you elaborate (maybe in a separate issue)?

anyzelman commented:
Attached this to milestone 0.9 btw (so blocked by 0.8)

GiovaGa committed:
Added dSB + random starting x and y + minor changes

In ising_machine_sb.hpp: Added dSB template parameter, changed some names to match reference

In test/smoke/ising_machine_sb: Added rand and zero options for input files + seed option

GiovaGa committed:
One less eWiseMul in kernel

GiovaGa committed:
Added wrappers for dSB and bSB

GiovaGa committed:
Removed mask and fold

GiovaGa committed:
Removed an mxv for dSB

GiovaGa committed:
Operator magic to avoid eWiseLambda

djelovina commented on include/graphblas/algorithms/ising_machine_sb.hpp:
```diff
@@ -50,31 +50,62 @@ namespace grb {
+			operator RetType() const { return val; }
+		};
+
+		// Custom type to trick assigning into doing hard clipping to [-1,1]
```
This looks nice!

GiovaGa committed:
Small improvements in ising_machine_sb
